src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -651,7 +651,33 @@ private BoundDecisionDag MakeBoundDecisionDag(SyntaxNode syntax, ImmutableArray<
 
             var rootDecisionDagNode = decisionDag.RootNode.Dag;
             RoslynDebug.Assert(rootDecisionDagNode != null);
-            return new BoundDecisionDag(rootDecisionDagNode.Syntax, rootDecisionDagNode);
+            var boundDecisionDag = new BoundDecisionDag(rootDecisionDagNode.Syntax, rootDecisionDagNode);
+#if DEBUG
+            int nextTempNumber = 0;
+            // Note that this leverages the `BoundDagEvaluation.GetHashCode()`
+            // to make "equivalent" evaluation nodes share an ID
+            var tempIdentifierMap = PooledDictionary<BoundDagEvaluation, int>.GetInstance();
+            int tempIdentifier(BoundDagEvaluation e)
```

Comment:
nit: I really have trouble with local functions in the middle of the logic. But we may be stuck here. Still, it may be better if this was just before the `#endif`.

Reply:
Fixed